python/amici/ode_export.py

```diff
@@ -1465,6 +1584,70 @@ def generate_basic_variables(self, *, from_sbml: bool = False) -> None:
 
         self._generate_symbol('x', from_sbml=from_sbml)
 
+    def get_events(self) -> None:
+        # We need to check the RHS for Heaviside functions (and later: events)
+        # track old all (unique) roots in roots
+        roots = []
+
+        def _dissect_expression(args):
+            root_fun_list = []
+            for arg in args:
+                if arg.func == sp.Heaviside:
+                    root_fun_list.append(arg.args[0])
+                elif arg.has(sp.Heaviside):
+                    root_fun_list.append(_dissect_expression(arg.args))
+            # flatten the expression
+            root_funs = []
+            for item in root_fun_list:
+                if type(item) is list:
+                    for subitem in item:
+                        root_funs.append(subitem)
+                else:
+                    root_funs.append(item)
+
+            return root_funs
+
+        def _make_unique(root_found, roots):
+            for root in roots:
+                if sp.simplify(root_found - root._value) == 0:
+                    return root._identifier, roots
+            else:
+                root_symstr = f'HeavisideFunction_{len(roots)}'
+                roots.append(Event(
+                    identifier=sp.Symbol(root_symstr),
+                    name=root_symstr,
+                    value=root_found,
+                    state_update=None,
+                    event_observable=None
+                ))
+                return roots[-1]._identifier, roots
+
+        def _extract_heavisides(dxdt, roots):
+            # expanding the rhs wll in general help to collect the same
+            # heaviside function
+            dt_expanded = dxdt.expand()
+            # track all the old Heaviside expressions in tmp_roots_old
+            # replace them later by the new expressions
+            heavisides = []
+            # run through the expression tree and get the roots
+            tmp_roots_old = _dissect_expression(dt_expanded.args)
+            for tmp_old in tmp_roots_old:
+                # we want unique identifiers for the roots
+                tmp_new, roots = _make_unique(tmp_old, roots)
+                _, roots = _make_unique(sp.sympify(-1 * tmp_old), roots)
+                heavisides.append((sp.Heaviside(tmp_old), tmp_new))
+
+            return roots, heavisides
+
+        for state in self._states:
+            roots, heavisides = _extract_heavisides(state._dt, roots)
+            for heaviside_sympy, heaviside_amici in heavisides:
+                state._dt = state._dt.subs(heaviside_sympy, heaviside_amici)
```

Comment:
Isn't it possible to already perform the substitution while performing `extract_heavisides`? Applying subs can be quite costly especially for larger models.

Reply:
It's already done state by state right after parsing, so the whole right hand side is not checked more often than necessary, I hope. Replacing right at checking for occurrence will not work well, as this will/must lead to redundantly saved roots. One could go for parsing string instead of symbols, but I'm somewhat sure that sympy does exactly this...
I tried to optimize it a little further by moving subs into the _extract_heavisides function, hence only those roots, which were found in a certain line of the rhs actually get replaced, and no others will be attempted (which will save time).
But beyond this, I think thereÄs no further optimization which will not break some functionality...